### PR TITLE
feat: wdpost: Parallel wdpost batch processing

### PR DESCRIPTION
**Pull request by magik6k**
## Related Issues
On top of https://github.com/filecoin-project/lotus/pull/9613

## Proposed Changes
* Process partition batches (PoSt messages) in parallel
* Add a parallelism limit config (default to 1 for no change in behavior)

## Additional Info
For whatever reason it was really bugging me that we weren't doing this, so now we do.

This should make WindowPoSt drastically faster for SPs with multiple partitions per deadline and make good use of multiple WindowPoSt workers. Also reorganizes / cleans up the runPostCycle code a little, which should make it easier to add other nice features like redundant compute.

SPs who have are submitting multiple PoSts per deadline can now compute them in parallel by turning a config knob:
```
[Proving]
  # Maximum number of partition batches to process in parellel. 0 = no limit.
  # 
  # A partition batch after being processed ends up as a PoSt message, so this limit can also be thought of as "how
  # many WindowPoSt messages to create in parallel".
  # 
  # Note that this is only really useful when multiple PoSt workers are connected and available. If there are no or
  # only one PoSt workers the effect will be limited, and performance may even degrade.
  #
  # type: int
  # env var: LOTUS_PROVING_MAXBATCHPARALLELISM
  MaxBatchParallelism = 10
```

TODO:
- [x] Write an itest which checks that multiple PoSt workers are used in parallel
- [x] Test with a big enough SP to see that this actually runs PoSt on multiple PoSt workers, and see if it actually is [partition-batches-per-deadline]x times faster

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] Tests exist for new functionality or change in behavior
- [x] CI is green

### CODE REVIEW

**Comment by magik6k** [comment]
Wrote an itest, and it's telling me that things don't quite work yet

File: storage/wdpost/wdpost_run.go [hunk 1]
```diff
@@ -300,205 +291,261 @@ func (s *WindowPoStScheduler) runPoStCycle(ctx context.Context, manual bool, di
...
+				return nil, xerrors.Errorf("received no proofs back from generate window post")
 			}
 
-			// Proof generation failed, so retry
-			log.Debugf("Proof generation failed, retry")
-			if len(ps) == 0 {
-				// If we didn't skip any new sectors, we failed
-				// for some other reason and we need to abort.
-				return nil, xerrors.Errorf("running window post failed: %w", err)
+			headTs, err := s.api.ChainHead(ctx)
```
[the remark this answers is not in the document]
Comment (magik6k) [reply storage/wdpost/wdpost_run.go:475]: Yeah, this was fine on master, but not here. I guess we can pull StateGetRandomnessFromBeacon out of the retry loop

**Comment by rjan90** [comment]
So running this on a local developer network with 4 partitions in a deadline + multiple windowPoSt-workers (3), it parallelizes the windowPoSt tasks across the workers, and runs them in parallel:

Settings on the lotus-miner (config.toml)
```
DisableBuiltinWindowPoSt = true
MaxBatchParallelism = 0
```

**PoSt-cycle starts:**

```
2023-08-04T07:21:35.096-0400    INFO    wdpost  wdpost/wdpost_run.go:306        starting PoSt cycle     {"cycle": "2023-08-04T07:21:35.094-0400", "manual": false, "ts": "[bafy2bzaceconpnuwmkkgwa6vctktli6o2sqpmmzxh75kqkqev37pcsuqh4jpu]", "deadline": 40, "batches": 1, "partitions": 4}
2023-08-04T07:21:35.101-0400    WARN    wdpost  wdpost/wdpost_run.go:249        Checked sectors {"checked": 2, "good": 2}
2023-08-04T07:21:35.105-0400    WARN    wdpost  wdpost/wdpost_run.go:249        Checked sectors {"checked": 2, "good": 2}
2023-08-04T07:21:35.110-0400    WARN    wdpost  wdpost/wdpost_run.go:249        Checked sectors {"checked": 2, "good": 2}
2023-08-04T07:21:35.114-0400    WARN    wdpost  wdpost/wdpost_run.go:249        Checked sectors {"checked": 2, "good": 2}
2023-08-04T07:21:35.115-0400    INFO    wdpost  wdpost/wdpost_run.go:459        running window post     {"batch": 0, "chain-random": "la80Y7bN+sJQwcp1qJemDX+rLIg7eXdkbaB+mj6kpyA=", "deadline": {"CurrentEpoch":3039,"PeriodStart":658,"Index":40,"Open":3058,"Close":3118,"Challenge":3038,"FaultCutoff":2988,"WPoStPeriodDeadlines":48,"WPoStProvingPeriod":2880,"WPoStChallengeWindow":60,"WPoStChallengeLookback":20,"FaultDeclarationCutoff":70}, "height": "3039", "skipped": 0, "partitions": 4}
2023-08-04T07:21:35.115-0400    INFO    advmgr  sealer/manager_post.go:137      generateWindowPoSt maxPartitionSize:2 partitionCount:4
2023-08-04T07:21:35.115-0400    INFO    advmgr  sealer/manager_post.go:219      generateWindowPost      {"index": 3}
2023-08-04T07:21:35.115-0400    INFO    advmgr  sealer/manager_post.go:219      generateWindowPost      {"index": 1}
2023-08-04T07:21:35.115-0400    INFO    advmgr  sealer/manager_post.go:219      generateWindowPost      {"index": 2}
2023-08-04T07:21:35.115-0400    INFO    advmgr  sealer/manager_post.go:219      generateWindowPost      {"index": 0}
2023-08-04T07:21:35.644-0400    WARN    advmgr  sealer/manager_post.go:234      generateWindowPost done {"index": 2, "skipped": 0, "took": "529.117966ms", "err": null}
2023-08-04T07:21:36.135-0400    WARN    advmgr  sealer/manager_post.go:234      generateWindowPost done {"index": 1, "skipped": 0, "took": "1.020679004s", "err": null}
2023-08-04T07:21:36.232-0400    INFO    advmgr  sealer/manager_post.go:24       GenerateWinningPoSt run at lotus-miner
2023-08-04T07:21:36.674-0400    WARN    advmgr  sealer/manager_post.go:234      generateWindowPost done {"index": 0, "skipped": 0, "took": "1.559765231s", "err": null}
2023-08-04T07:21:37.181-0400    WARN    advmgr  sealer/manager_post.go:234      generateWindowPost done {"index": 3, "skipped": 0, "took": "2.066316791s", "err": null}
2023-08-04T07:21:37.181-0400    INFO    wdpost  wdpost/wdpost_run.go:480        computing window post   {"batch": 0, "batch": 0, "elapsed": 2.06646504, "skip": 0, "err": null}
2023-08-04T07:21:37.195-0400    INFO    wdpost  wdpost/wdpost_run.go:275        post cycle done {"cycle": "2023-08-04T07:21:35.094-0400", "took": 2.100437836, "postCount": 1}
```

**On windowPoSt-worker 1:**

```
2023-08-04T07:21:35.127 INFO filecoin_proofs::api::window_post > generate_single_window_post_with_vanilla:start
2023-08-04T07:21:35.127 INFO filecoin_proofs::caches > trying parameters memory cache for: Window_POST[2048]
2023-08-04T07:21:35.127 INFO filecoin_proofs::caches > found params in memory cache for Window_POST[2048]
2023-08-04T07:21:35.127 INFO filecoin_proofs::api::post_util > single_partition_vanilla_proofs:start
2023-08-04T07:21:35.127 INFO filecoin_proofs::api::post_util > single_partition_vanilla_proofs:finish
2023-08-04T07:21:35.127 INFO storage_proofs_core::compound_proof > snark_proof:start
2023-08-04T07:21:35.127 INFO bellperson::groth16::prover > Bellperson 0.24.1 is being used!
2023-08-04T07:21:35.154 INFO bellperson::groth16::prover > synthesis time: 26.988026ms
2023-08-04T07:21:35.154 INFO bellperson::groth16::prover > starting proof timer
-------
2023-08-04T07:21:35.643 INFO bellperson::groth16::prover > prover time: 489.045038ms
2023-08-04T07:21:35.644 INFO storage_proofs_core::compound_proof > snark_proof:finish
2023-08-04T07:21:35.644 INFO filecoin_proofs::api::window_post > generate_single_window_post_with_vanilla:finish
```

**On windowPoSt-worker 2:**

```
023-08-04T07:21:35.128 INFO filecoin_proofs::api::window_post > generate_single_window_post_with_vanilla:start
2023-08-04T07:21:35.128 INFO filecoin_proofs::caches > trying parameters memory cache for: Window_POST[2048]
2023-08-04T07:21:35.128 INFO filecoin_proofs::caches > found params in memory cache for Window_POST[2048]
2023-08-04T07:21:35.128 INFO filecoin_proofs::api::post_util > single_partition_vanilla_proofs:start
2023-08-04T07:21:35.128 INFO filecoin_proofs::api::post_util > single_partition_vanilla_proofs:finish
2023-08-04T07:21:35.130 INFO storage_proofs_core::compound_proof > snark_proof:start
2023-08-04T07:21:35.130 INFO bellperson::groth16::prover > Bellperson 0.24.1 is being used!
2023-08-04T07:21:35.171 INFO bellperson::groth16::prover > synthesis time: 41.276967ms
2023-08-04T07:21:35.171 INFO bellperson::groth16::prover > starting proof timer
------
2023-08-04T07:21:36.132 INFO bellperson::groth16::prover > prover time: 960.406158ms
2023-08-04T07:21:36.135 INFO storage_proofs_core::compound_proof > snark_proof:finish
2023-08-04T07:21:36.135 INFO filecoin_proofs::api::window_post > generate_single_window_post_with_vanilla:finish
```

**On windowPoSt-worker 3 (which did proving for two partitions, ref double logs):**

```
2023-08-04T07:21:35.129 INFO filecoin_proofs::api::window_post > generate_single_window_post_with_vanilla:start
2023-08-04T07:21:35.129 INFO filecoin_proofs::caches > trying parameters memory cache for: Window_POST[2048]
2023-08-04T07:21:35.129 INFO filecoin_proofs::caches > found params in memory cache for Window_POST[2048]
2023-08-04T07:21:35.129 INFO filecoin_proofs::api::post_util > single_partition_vanilla_proofs:start
2023-08-04T07:21:35.129 INFO filecoin_proofs::api::post_util > single_partition_vanilla_proofs:finish
2023-08-04T07:21:35.129 INFO filecoin_proofs::api::window_post > generate_single_window_post_with_vanilla:start
2023-08-04T07:21:35.129 INFO filecoin_proofs::caches > trying parameters memory cache for: Window_POST[2048]
2023-08-04T07:21:35.129 INFO filecoin_proofs::caches > found params in memory cache for Window_POST[2048]
2023-08-04T07:21:35.129 INFO filecoin_proofs::api::post_util > single_partition_vanilla_proofs:start
2023-08-04T07:21:35.129 INFO filecoin_proofs::api::post_util > single_partition_vanilla_proofs:finish
2023-08-04T07:21:35.130 INFO storage_proofs_core::compound_proof > snark_proof:start
2023-08-04T07:21:35.130 INFO storage_proofs_core::compound_proof > snark_proof:start
2023-08-04T07:21:35.130 INFO bellperson::groth16::prover > Bellperson 0.24.1 is being used!
2023-08-04T07:21:35.130 INFO bellperson::groth16::prover > Bellperson 0.24.1 is being used!
2023-08-04T07:21:35.173 INFO bellperson::groth16::prover > synthesis time: 43.077058ms
2023-08-04T07:21:35.173 INFO bellperson::groth16::prover > starting proof timer
2023-08-04T07:21:35.176 INFO bellperson::groth16::prover > synthesis time: 45.710681ms
2023-08-04T07:21:35.176 INFO bellperson::groth16::prover > starting proof timer
----------
2023-08-04T07:21:36.672 INFO bellperson::groth16::prover > prover time: 1.498767489s
2023-08-04T07:21:36.674 INFO storage_proofs_core::compound_proof > snark_proof:finish
2023-08-04T07:21:36.674 INFO filecoin_proofs::api::window_post > generate_single_window_post_with_vanilla:finish
----------
2023-08-04T07:21:37.180 INFO bellperson::groth16::prover > prover time: 2.00414668s
2023-08-04T07:21:37.181 INFO storage_proofs_core::compound_proof > snark_proof:finish
2023-08-04T07:21:37.181 INFO filecoin_proofs::api::window_post > generate_single_window_post_with_vanilla:finish
``

File: documentation/en/default-lotus-miner-config.toml [hunk 2]
```diff
@@ -466,6 +468,22 @@
   # env var: LOTUS_PROVING_SINGLERECOVERINGPARTITIONPERPOSTMESSAGE
   #SingleRecoveringPartitionPerPostMessage = false
 
+  # Maximum number of partition batches to process in parellel. 0 = no limit.
+  # 
+  # A partition batch after being processed ends up as a PoSt message, so this limit can also be thought of as "how
+  # many WindowPoSt messages to create in parallel".
```
Comment (shrenujbansal) [inline documentation/en/default-lotus-miner-config.toml:474]: Little confused here - I thought we were processing multiple partitions in the same WdPost msg in parallel

File: itests/worker_test.go [hunk 3]
```diff
@@ -438,6 +441,129 @@ func TestWindowPostWorkerDisconnected(t *testing.T) {
...
+
+func TestWindowPostWorkerParallel(t *testing.T) {
+	doTest := func(maxBatch, maxPart int) func(t *testing.T) {
+		return func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			_ = logging.SetLogLevel("storageminer", "INFO")
+
+			sectors := 2 * 48 * 2
```
Comment (shrenujbansal) [inline itests/worker_test.go:466]: Could you add a comment saying how you came up with this number?

Comment (rjan90) [reply itests/worker_test.go:466]: This number is coming from that a local network has a max of 2 sectors per partition. So 2 sectors * 48 windows * 2 partitions, to be able to test that windowPoSts are processed in parallel. I can that comment if that is okay for you @magik6k ?

File: documentation/en/default-lotus-miner-config.toml (hunk 2 above)
Comment (magik6k) [reply documentation/en/default-lotus-miner-config.toml:474]: No, it's exactly as this comment says, the loop in `runPoStCycle` starts a goroutine per batch, and a batch can contain more than one partition. We're not able to compute partitions within one batch in parallel due to how sector skipping / challenge generation used to work in proofs. The new sector challenge generation has now landed in the network, but we still didn't update the windowpost logic to do the simpler skipping (and that's not something I want to do as part of this PR)

**Review by snadrus — CHANGES_REQUESTED** [review]

**Comment by magik6k** [comment]
This can now be done with Curio

**Pull request closed without merging**